File: packages/ts-sdk/src/core/distances.ts
import {
  absoluteInteger,
  cloneNumberArray,
  sortAscending
} from "./internal";
import { IntervalVector } from "./intervalVector";
import { PositionVector } from "./positionVector";

type DistanceInput = readonly number[] | PositionVector | IntervalVector;

function transformationSteps(start: readonly number[], end: readonly number[]): Array<[number, [number, number]]> {
  var steps: Array<[number, [number, number]]> = [];
  var minLength = start.length < end.length ? start.length : end.length;
  var index = 0;

  for (index = 0; index < minLength; index += 1) {
    var differenceValue = end[index] - start[index];
    if (differenceValue !== 0) {
      steps.push([0, [index, differenceValue]]);
      var transformed = cloneNumberArray(start);
      transformed[index] = transformed[index] + differenceValue;
      return steps.concat(transformationSteps(transformed, end));
    }
  }

  if (end.length > start.length) {
    for (index = minLength; index < end.length; index += 1) {
      steps.push([1, [start.length + (index - minLength), end[index]]]);
    }
  }

  if (end.length < start.length) {
    for (index = minLength; index < start.length; index += 1) {
      steps.push([2, [index, start[index]]]);
    }
  }

  return steps;
}

function distanceSource(input: DistanceInput): number[] {
  if (input instanceof PositionVector || input instanceof IntervalVector) {
    return input.toArray();
  }
  return cloneNumberArray(input);
}

export function euclideanDistance(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var length = leftData.length < rightData.length ? leftData.length : rightData.length;
  var total = 0;

  for (var index = 0; index < length; index += 1) {
    var differenceValue = leftData[index] - rightData[index];
    total += differenceValue * differenceValue;
  }

  return Math.sqrt(total);
}

export function editDistance(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var rows = leftData.length + 1;
  var columns = rightData.length + 1;
  var dp: number[][] = [];
  var row = 0;
  var column = 0;

  for (row = 0; row < rows; row += 1) {
    dp[row] = [];
    dp[row][0] = row;
  }
  for (column = 0; column < columns; column += 1) {
    dp[0][column] = column;
  }

  for (row = 1; row < rows; row += 1) {
    for (column = 1; column < columns; column += 1) {
      if (leftData[row - 1] === rightData[column - 1]) {
        dp[row][column] = dp[row - 1][column - 1];
      } else {
        var insertion = dp[row][column - 1];
        var deletion = dp[row - 1][column];
        var substitution = dp[row - 1][column - 1];
        var minimum = insertion < deletion ? insertion : deletion;
        minimum = minimum < substitution ? minimum : substitution;
        dp[row][column] = 1 + minimum;
      }
    }
  }

  return dp[rows - 1][columns - 1];
}

export function hammingDistance(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var length = leftData.length < rightData.length ? leftData.length : rightData.length;
  var distance = 0;

  for (var index = 0; index < length; index += 1) {
    if (leftData[index] !== rightData[index]) {
      distance += 1;
    }
  }

  return distance;
}

export function manhattanDistance(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var length = leftData.length < rightData.length ? leftData.length : rightData.length;
  var sum = 0;

  for (var index = 0; index < length; index += 1) {
    sum += absoluteInteger(leftData[index] - rightData[index]);
  }

  return sum;
}

export function difference(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var length = leftData.length < rightData.length ? leftData.length : rightData.length;
  var total = 0;

  for (var index = 0; index < length; index += 1) {
    total += leftData[index] - rightData[index];
  }

  return total;
}

export function variationDistance(left: readonly number[] | PositionVector, right: readonly number[] | PositionVector): number {
  var leftData = left instanceof PositionVector ? left.toArray() : cloneNumberArray(left);
  var rightData = right instanceof PositionVector ? right.toArray() : cloneNumberArray(right);

  if (leftData.length === 0 && rightData.length === 0) {
    return 0;
  }
  if (leftData.length === 0 || rightData.length === 0) {
    throw new Error("Empty vector, distance calculation is impossible.");
  }

  var sortedLeft = sortAscending(leftData);
  var sortedRight = sortAscending(rightData);
  var minLength = sortedLeft.length < sortedRight.length ? sortedLeft.length : sortedRight.length;
  var maxLength = sortedLeft.length > sortedRight.length ? sortedLeft.length : sortedRight.length;
  var total = 0;
  var index = 0;

  for (index = 0; index < minLength; index += 1) {
    total += absoluteInteger(sortedLeft[index] - sortedRight[index]);
  }

  var longer = sortedLeft.length >= sortedRight.length ? sortedLeft : sortedRight;
  var shorter = sortedLeft.length >= sortedRight.length ? sortedRight : sortedLeft;

  for (index = minLength; index < maxLength; index += 1) {
    var extra = longer[index];
    var minimum = Number.POSITIVE_INFINITY;
    var maximum = Number.NEGATIVE_INFINITY;
    for (var shortIndex = 0; shortIndex < shorter.length; shortIndex += 1) {
      var currentDistance = absoluteInteger(extra - shorter[shortIndex]);
      if (currentDistance < minimum) {
        minimum = currentDistance;
      }
      if (currentDistance > maximum) {
        maximum = currentDistance;
      }
    }
    total += (minimum + maximum) / 2;
  }

  return total;
}

export function weightedTransformationDistance(left: DistanceInput, right: DistanceInput): number {
  var leftData = distanceSource(left);
  var rightData = distanceSource(right);
  var steps = transformationSteps(leftData, rightData);
  var distance = 0;

  for (var index = 0; index < steps.length; index += 1) {
    distance += absoluteInteger(steps[index][1][1]);
  }

  return distance;
}

File: packages/dotnet-sdk/src/Distances.cs
namespace Musicpp;

public static class Distances
{
    public static List<double> Normalize(IReadOnlyList<int> input)
    {
        var sum = input.Sum(x => (double)x);
        if (sum == 0)
            throw new ArgumentException("Sum of vector elements is zero, cannot normalize");
        return input.Select(v => v / sum).ToList();
    }

    public static List<double> ComputeCdf(List<double> pdf)
    {
        var cdf = new List<double>(pdf.Count);
        double acc = 0;
        foreach (var p in pdf)
        {
            acc += p;
            cdf.Add(acc);
        }
        return cdf;
    }

    public static double EuclideanDistance(IReadOnlyList<int> v1, IReadOnlyList<int> v2)
    {
        var length = Math.Min(v1.Count, v2.Count);
        double sum = 0;
        for (var i = 0; i < length; i++)
        {
            var d = v1[i] - v2[i];
            sum += d * d;
        }
        return Math.Sqrt(sum);
    }

    public static int EditDistance(List<int> v1, List<int> v2)
    {
        var n = v1.Count;
        var m = v2.Count;
        var dp = new int[n + 1, m + 1];
        for (var i = 0; i <= n; i++) dp[i, 0] = i;
        for (var j = 0; j <= m; j++) dp[0, j] = j;
        for (var i = 1; i <= n; i++)
            for (var j = 1; j <= m; j++)
            {
                if (v1[i - 1] == v2[j - 1])
                    dp[i, j] = dp[i - 1, j - 1];
                else
                    dp[i, j] = 1 + Math.Min(Math.Min(dp[i - 1, j], dp[i, j - 1]), dp[i - 1, j - 1]);
            }
        return dp[n, m];
    }

    public static int HammingDistance(List<int> v1, List<int> v2)
    {
        var length = Math.Min(v1.Count, v2.Count);
        var d = 0;
        for (var i = 0; i < length; i++)
            if (v1[i] != v2[i])
                d++;
        return d;
    }

    public static int ManhattanDistance(List<int> v1, List<int> v2)
    {
        var length = Math.Min(v1.Count, v2.Count);
        var sum = 0;
        for (var i = 0; i < length; i++)
            sum += Math.Abs(v1[i] - v2[i]);
        return sum;
    }

    public static int Difference(List<int> v1, List<int> v2)
    {
        var length = Math.Min(v1.Count, v2.Count);
        var diff = 0;
        for (var i = 0; i < length; i++)
            diff += v1[i] - v2[i];
        return diff;
    }

    public static double VariationDistance(List<int> a, List<int> b)
    {
        if (a.Count == 0 && b.Count == 0)
            return 0;
        if (a.Count == 0 || b.Count == 0)
            throw new ArgumentException("Empty vector, distance calculation is impossible.");
        a = a.OrderBy(x => x).ToList();
        b = b.OrderBy(x => x).ToList();
        var minLen = Math.Min(a.Count, b.Count);
        var maxLen = Math.Max(a.Count, b.Count);
        double total = 0;
        for (var i = 0; i < minLen; i++)
            total += Math.Abs(a[i] - b[i]);
        var longer = a.Count >= b.Count ? a : b;
        var shorter = a.Count >= b.Count ? b : a;
        for (var i = minLen; i < maxLen; i++)
        {
            var extra = longer[i];
            var minDist = int.MaxValue;
            var maxDist = int.MinValue;
            foreach (var x in shorter)
            {
                var d = Math.Abs(extra - x);
                minDist = Math.Min(minDist, d);
                maxDist = Math.Max(maxDist, d);
            }
            total += (minDist + maxDist) / 2.0;
        }
        return total;
    }

    public static List<int> GeneralizedNeoRiemann(IReadOnlyList<int> input, int position, int shift)
    {
        var output = input.ToList();
        if (position >= 0 && position < output.Count)
            output[position] += shift;
        return output;
    }

    public static List<(int Type, int Pos, int Val)> TransformationSteps(IReadOnlyList<int> start, IReadOnlyList<int> end)
    {
        var steps = new List<(int, int, int)>();
        var startLength = start.Count;
        var endLength = end.Count;
        var minLength = Math.Min(startLength, endLength);
        var addedPosition = startLength;
        for (var i = 0; i < minLength; i++)
        {
            var diff = end[i] - start[i];
            if (diff != 0)
            {
                steps.Add((0, i, diff));
                var transformed = GeneralizedNeoRiemann(start.ToList(), i, diff);
                steps.AddRange(TransformationSteps(transformed, end));
                return steps;
            }
        }
        if (endLength > startLength)
        {
            for (var i = minLength; i < endLength; i++)
                steps.Add((1, addedPosition++, end[i]));
        }
        if (endLength < startLength)
        {
            for (var i = minLength; i < startLength; i++)
                steps.Add((2, i, start[i]));
        }
        return steps;
    }

    public static int WeightedTransformationDistance(List<int> start, List<int> end)
    {
        var steps = TransformationSteps(start, end);
        return steps.Sum(s => Math.Abs(s.Val));
    }

    public static double EuclideanDistance(PositionVector a, PositionVector b) =>
        EuclideanDistance(a.MutableData, b.MutableData);

    public static int ManhattanDistance(PositionVector a, PositionVector b) =>
        ManhattanDistance(a.MutableData, b.MutableData);

    public static int EditDistance(PositionVector a, PositionVector b) =>
        EditDistance(a.MutableData.ToList(), b.MutableData.ToList());

    public static int WeightedTransformationDistance(PositionVector a, PositionVector b) =>
        WeightedTransformationDistance(a.MutableData.ToList(), b.MutableData.ToList());

    public static int Difference(PositionVector a, PositionVector b) =>
        Difference(a.MutableData, b.MutableData);

    public static int HammingDistance(PositionVector a, PositionVector b) =>
        HammingDistance(a.MutableData, b.MutableData);

    public static int Difference(IntervalVector a, IntervalVector b) =>
        Difference(a.MutableData, b.MutableData);

    public static int HammingDistance(IntervalVector a, IntervalVector b) =>
        HammingDistance(a.MutableData, b.MutableData);

    public static int ManhattanDistance(IntervalVector a, IntervalVector b) =>
        ManhattanDistance(a.MutableData, b.MutableData);

    public static double EuclideanDistance(IntervalVector a, IntervalVector b) =>
        EuclideanDistance(a.MutableData, b.MutableData);

    public static int EditDistance(IntervalVector a, IntervalVector b) =>
        EditDistance(a.MutableData.ToList(), b.MutableData.ToList());

    public static int WeightedTransformationDistance(IntervalVector a, IntervalVector b) =>
        WeightedTransformationDistance(a.MutableData.ToList(), b.MutableData.ToList());

    public static double VariationDistance(PositionVector a, PositionVector b) =>
        VariationDistance(a.MutableData.ToList(), b.MutableData.ToList());
}
